src/network/networkd-ndisc.c

```diff
@@ -1334,6 +1335,93 @@ static int ndisc_router_process_pref64(Link *link, sd_ndisc_router *rt) {
         return 0;
 }
 
+static NDiscDNR* ndisc_dnr_free(NDiscDNR *x) {
+        if (!x)
+                return NULL;
+
+        sd_dns_resolver_done(&x->resolver);
+        return mfree(x);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(NDiscDNR*, ndisc_dnr_free);
+
+static int ndisc_dnr_compare_func(const NDiscDNR *a, const NDiscDNR *b) {
+        return strcmp_ptr(a->resolver.auth_name, b->resolver.auth_name);
+}
+
+static void ndisc_dnr_hash_func(const NDiscDNR *x, struct siphash *state) {
+        assert(x);
+
+        siphash24_compress_typesafe(x->resolver.transports, state);
+        siphash24_compress_typesafe(x->resolver.port, state);
+
+        siphash24_compress_string(x->resolver.auth_name, state);
+        siphash24_compress_string(x->resolver.dohpath, state);
+
+        siphash24_compress_typesafe(x->resolver.family, state);
+        FOREACH_ARRAY(addr, x->resolver.addrs, x->resolver.n_addrs)
+                siphash24_compress_typesafe(*addr, state);
+}
+
+DEFINE_PRIVATE_HASH_OPS_WITH_KEY_DESTRUCTOR(
+                ndisc_dnr_hash_ops,
+                NDiscDNR,
+                ndisc_dnr_hash_func,
+                ndisc_dnr_compare_func,
+                ndisc_dnr_free);
+
+static int ndisc_router_process_encrypted_dns(Link *link, sd_ndisc_router *rt) {
+        int r = 0;
+
+        assert(link);
+        assert(link->network);
+        assert(rt);
+
+        struct in6_addr router;
+        _cleanup_(sd_dns_resolver_done) sd_dns_resolver res = {};
+        _cleanup_(ndisc_dnr_freep) NDiscDNR *new_entry = NULL;
+
+        if (!network_ndisc_use_dnr(link->network))
+                return 0;
+
+        r = sd_ndisc_router_get_sender_address(rt, &router);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get gateway address from RA: %m");
+
+        usec_t lifetime_usec;
+        r = sd_ndisc_router_encrypted_dns_get_lifetime_timestamp(rt, CLOCK_BOOTTIME, &lifetime_usec);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get lifetime of RA message: %m");
+
+        r = sd_ndisc_router_encrypted_dns_get_dnr(rt, &res);
```

Comment:
```sh
=================================================================
==187==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x608000001bf8 at pc 0x000000a926e9 bp 0x7fffb5b57c70 sp 0x7fffb5b57c68
READ of size 2 at 0x608000001bf8 thread T0
    #0 0xa926e8 in unaligned_read_be16 /home/vagrant/systemd/build/../src/basic/unaligned.h:14:27
    #1 0xa912fd in sd_ndisc_router_encrypted_dns_get_dnr /home/vagrant/systemd/build/../src/libsystemd-network/sd-ndisc-router.c:264:23
    #2 0x8792b5 in ndisc_router_process_encrypted_dns /home/vagrant/systemd/build/../src/network/networkd-ndisc.c:1396:13
    #3 0x860430 in ndisc_router_process_options /home/vagrant/systemd/build/../src/network/networkd-ndisc.c:1477:29
    #4 0x857e3a in ndisc_router_handler /home/vagrant/systemd/build/../src/network/networkd-ndisc.c:1762:13
    #5 0x855cff in ndisc_handler /home/vagrant/systemd/build/../src/network/networkd-ndisc.c:1792:21
    #6 0xa85891 in ndisc_callback /home/vagrant/systemd/build/../src/libsystemd-network/sd-ndisc.c:42:9
    #7 0xa84c1b in ndisc_handle_router /home/vagrant/systemd/build/../src/libsystemd-network/sd-ndisc.c:223:9
    #8 0xa834db in ndisc_recv /home/vagrant/systemd/build/../src/libsystemd-network/sd-ndisc.c:261:24
    #9 0x7fdb98d5281c in source_dispatch /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4187:21
    #10 0x7fdb98d50844 in sd_event_dispatch /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4808:21
    #11 0x7fdb98d5633d in sd_event_run /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4869:21
    #12 0x7fdb98d572c9 in sd_event_loop /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4891:21
    #13 0x4170ed in run /home/vagrant/systemd/build/../src/network/networkd.c:109:13
    #14 0x41417d in main /home/vagrant/systemd/build/../src/network/networkd.c:116:1
    #15 0x7fdb97249b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #16 0x7fdb97249c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #17 0x412b94 in _start (/usr/lib/systemd/systemd-networkd+0x412b94) (BuildId: 9ab9e1ab528679dc87e353628a42ba0f384f6644)
0x608000001bf8 is located 0 bytes after 88-byte region [0x608000001ba0,0x608000001bf8)
allocated by thread T0 here:
    #0 0x7fdb996eb1bc in __interceptor_calloc (/usr/lib64/clang/16/lib/libclang_rt.asan-x86_64.so+0xeb1bc) (BuildId: 2a39bb4aa6d13ff866a4a47997cefdbe31a0350f)
    #1 0xad5029 in icmp6_packet_new /home/vagrant/systemd/build/../src/libsystemd-network/icmp6-packet.c:20:13
    #2 0xad3e77 in icmp6_packet_receive /home/vagrant/systemd/build/../src/libsystemd-network/icmp6-packet.c:95:13
    #3 0xa8254e in ndisc_recv /home/vagrant/systemd/build/../src/libsystemd-network/sd-ndisc.c:235:13
    #4 0x7fdb98d5281c in source_dispatch /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4187:21
    #5 0x7fdb98d50844 in sd_event_dispatch /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4808:21
    #6 0x7fdb98d5633d in sd_event_run /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4869:21
    #7 0x7fdb98d572c9 in sd_event_loop /home/vagrant/systemd/build/../src/libsystemd/sd-event/sd-event.c:4891:21
    #8 0x4170ed in run /home/vagrant/systemd/build/../src/network/networkd.c:109:13
    #9 0x41417d in main /home/vagrant/systemd/build/../src/network/networkd.c:116:1
    #10 0x7fdb97249b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #11 0x7fdb97249c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #12 0x412b94 in _start (/usr/lib/systemd/systemd-networkd+0x412b94) (BuildId: 9ab9e1ab528679dc87e353628a42ba0f384f6644)
SUMMARY: AddressSanitizer: heap-buffer-overflow /home/vagrant/systemd/build/../src/basic/unaligned.h:14:27 in unaligned_read_be16
Shadow bytes around the buggy address:
  0x608000001900: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x608000001980: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x608000001a00: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x608000001a80: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x608000001b00: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
=>0x608000001b80: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 00[fa]
  0x608000001c00: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 00 00
  0x608000001c80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x608000001d00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x608000001d80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x608000001e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==187==ABORTING
```
The option was
```
          "icmpv6.opt": {
            "icmpv6.opt.type": "144",
            "icmpv6.opt.length": "1",
            "icmpv6.data": "00:00:00:00:00:00"
          }
```

Reply:
Oh, I should have read yuwata's patch more carefully. I have some more changes to make... the current version is pretty borked I guess.